src/main/kotlin/org/rust/lang/core/types/analysis/mutability/UnusedMutability.kt

```diff
@@ -0,0 +1,186 @@
+        mutableDeclarations.add(ProblematicDeclaration(binding, kind))
+    }
+
+    fun check(): UnusedMutability? {
+        val gatherMutabilityContext = GatherUsageContext(MutabilityData())
+        val mutabilityData = gatherMutabilityContext.gather(body, implLookup, inference)
+        val flowedMutability = FlowedMutabilityData.buildFor(this, mutabilityData, cfg)
+        flowedMutability.check()
+        return UnusedMutability(mutableDeclarations)
+    }
+
+    companion object {
+        fun buildFor(owner: RsInferenceContextOwner): UnusedMutabilityContext? {
+            val body = owner.body as? RsBlock ?: return null
+            val cfg = owner.controlFlowGraph ?: return null
+            return UnusedMutabilityContext(owner.inference, body, cfg)
+        }
+    }
+}
+
+object MutabilityDataFlowOperator : DataFlowOperator {
+    override fun join(succ: Int, pred: Int): Int = succ or pred     // liveness from both preds are in scope
+    override val initialValue: Boolean get() = false                // does not need mutability by default
+}
+
+private typealias MutabilityDataFlow = DataFlowContext<MutabilityDataFlowOperator>
+
+private class FlowedMutabilityData(
+    private val ctx: UnusedMutabilityContext,
+    private val mutabilityData: UsageAnalysisData<MutabilityDataFlow>,
+    private val dfcxLivePaths: MutabilityDataFlow
+) {
+    private fun isImmutable(usagePath: UsagePath): Boolean {
+        var isImmutable = true
+        dfcxLivePaths.eachBitOnEntry(usagePath.declaration) { index ->
+            val path = mutabilityData.paths[index]
+            // mutable usage of `a`
+            if (usagePath == path) {
+                isImmutable = false
+            } else {
+                // declaration/assignment of `a`, use of `a.b.c`
+                val isEachExtensionDead = mutabilityData.eachBasePath(path) { it != usagePath }
+                if (!isEachExtensionDead) isImmutable = false
+            }
+            isImmutable
+        }
+        return isImmutable
+    }
+
+    fun check() {
+        for (path in mutabilityData.paths) {
+            if (path is UsagePath.Base && path.declaration.isMutable() && isImmutable(path)) {
+                ctx.reportNeedlesslyMutableDeclaration(path.declaration, path.declarationKind)
+            }
+        }
+    }
+
+    companion object {
+        fun buildFor(ctx: UnusedMutabilityContext, mutabilityData: UsageAnalysisData<MutabilityDataFlow>,
+                     cfg: ControlFlowGraph): FlowedMutabilityData {
+            val dfcxLivePaths = DataFlowContext(cfg, MutabilityDataFlowOperator, mutabilityData.paths.size, FlowDirection.Backward)
+
+            mutabilityData.addGenKills(dfcxLivePaths)
+            dfcxLivePaths.propagate()
+
+            return FlowedMutabilityData(ctx, mutabilityData, dfcxLivePaths)
+        }
+    }
+}
+
+private fun isMutableContext(element: RsElement): Boolean {
```

Comment:
Some of the logic looks duplicated because of `ExprUseWalker`. For example, the `RsCondition` case may be covered by `ExprUseWalker#walkCondition` which calls `ExprUseWalker#walkPat`. Please consider reusing `ExprUseWalker#walkPat`: this method is aimed to be general-purpose to analyze any pattern bindings.